src/gd_avif.c

```diff
@@ -0,0 +1,663 @@
+
+	if (!ctx)
+		return 0;
+
+	im = gdImageCreateFromAvifCtx(ctx);
+	ctx->gd_free(ctx);
+
+	return im;
+}
+
+/*
+	Function: gdImageCreateFromAvifCtx
+
+		See <gdImageCreateFromAvif>.
+		
+		Additional details: the AVIF library comes with functions to create an IO object from
+		a file and from a memory pointer. Of course, it doesn't have a way to create an IO object
+		from a gdIOCtx. So, here, we use our own helper function, <createAvifIOfromCtx>.
+
+		Otherwise, we create the image by calling AVIF library functions in order:
+		* avifDecoderCreate(), to create the decoder
+		* avifDecoderSetIO(), to tell libavif how to read from our data structure
+		* avifDecoderParse(), to parse the image
+		* avifDecoderNextImage(), to read the first image from the decoder
+		* avifRGBImageSetDefaults(), to create the avifRGBImage
+		* avifRGBImageAllocatePixels(), to allocate memory for the pixels
+		* avifImageYUVToRGB(), to convert YUV to RGB
+		
+		Finally, we create a new gd image and copy over the pixel data.
+	
+	Parameters:
+
+		ctx							- a gdIOCtx struct
+*/
+BGD_DECLARE(gdImagePtr) gdImageCreateFromAvifCtx (gdIOCtx *ctx)
+{
+	int x, y;
+	uint8_t *p8; 
+	uint16_t *p16;
+	gdImage *im = NULL;
+	avifResult result;
+	avifIO *io;
+	avifDecoder *decoder;
+	avifRGBImage rgb;
+
+	decoder = avifDecoderCreate();
+
+	io = createAvifIOFromCtx(ctx);
+	if (!io) {
+		gd_error("avif error: Could not allocate memory");
+		goto cleanup;		
+	}
+
+	avifDecoderSetIO(decoder, io);
+
+	result = avifDecoderParse(decoder);
+	if (isAvifError(result, "Could not parse image")) 
+		goto cleanup;
+
+	// insert sRGB check
+
+	// Note again that, for an image sequence, we read only the first image, ignoring the rest.
+	result = avifDecoderNextImage(decoder);
+	if (isAvifError(result, "Could not decode image")) 
+		goto cleanup;
+
+	// Set up the avifRGBImage, and convert it from YUV to RGB.
+	avifRGBImageSetDefaults(&rgb, decoder->image);
+	avifRGBImageAllocatePixels(&rgb);
+
+	result = avifImageYUVToRGB(decoder->image, &rgb);
+	if (isAvifError(result, "gd-avif error: Conversion from YUV to RGB failed")) 
+		goto cleanup;
+
+	im = gdImageCreateTrueColor(decoder->image->width, decoder->image->height);
+	if (!im) {
+		gd_error("avif error: Could not create GD truecolor image");
+		goto cleanup;
+	}
+
+// Read the pixels from the AVIF image and copy them into the GD image.
+// Pixel depth can be 8, 10, or 12 bits. But if depth>8, pixels are 16-bit.
+// Since avifDecoderNextImage() has succeeded, we know we've got a valid pixel depth.
+
+	if (rgb.depth == 8) {
+		for (y = 0, p8 = rgb.pixels; y < decoder->image->height; y++) {
+			for (x = 0; x < decoder->image->width; x++) {
+				uint8_t r = *(p8++);
+				uint8_t g = *(p8++);
+				uint8_t b = *(p8++);
+				uint8_t a = *(p8++);
+				im->tpixels[y][x] = gdTrueColorAlpha(r, g, b, a);
+			}
+		}
+		
+	} else {
+		for (y = 0, p16 = (uint16_t *) rgb.pixels; y < decoder->image->height; y++) {
```

Comment:
i guess you copied it from gd_webp, but move the `p16` assignment out of the for loop init.  it's unnecessary and doesn't improve readability.